django/contrib/auth/hashers.py

```diff
@@ -512,6 +512,88 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
     digest = None
 
 
+class ScryptPasswordHasher(BasePasswordHasher):
+    """
+    Secure password hashing using the Scrypt algorithm
+
+    The result is a 64 byte binary string.
+    """
+    algorithm = 'scrypt'
+    block_size = 8
+    maxmem = 0
+    parallelism = 1
+    work_factor = 2 ** 14
+
+    def encode(self, password, salt, n=None, r=None):
+        assert password is not None
+        assert salt and '$' not in salt
+        n = n or self.work_factor
+        r = r or self.block_size
+        hash_ = hashlib.scrypt(
+            password.encode(),
+            salt=salt.encode(),
+            n=n,
+            r=r,
+            p=self.parallelism,
+            maxmem=self.maxmem,
+            dklen=64,
+        )
+        hash_ = base64.b64encode(hash_).decode('ascii').strip()
+        return '%s$%d$%s$%s$%d$%d' % (
+            self.algorithm, n, salt, hash_, r, self.parallelism,
+        )
+
+    def decode(self, encoded):
+        algorithm, work_factor, salt, hash_, block_size, parallelism = encoded.split('$', 6)
+        assert algorithm == self.algorithm
+        return {
+            'algorithm': algorithm,
+            'hash': hash_,
+            'work_factor': int(work_factor),
+            'salt': salt,
+            'block_size': int(block_size),
+            'parallelism': int(parallelism),
+        }
+
+    def verify(self, password, encoded):
+        decoded = self.decode(encoded)
+        encoded_2 = self.encode(
+            password, decoded['salt'], decoded['work_factor'],
+            decoded['block_size']
+        )
+        return constant_time_compare(encoded, encoded_2)
+
+    def safe_summary(self, encoded):
+        decoded = self.decode(encoded)
+        return {
+            _('algorithm'): decoded['algorithm'],
+            _('work factor'): decoded['work_factor'],
+            _('block size'): decoded['block_size'],
+            _('parallelism'): decoded['parallelism'],
+            _('salt'): mask_hash(decoded['salt']),
+            _('hash'): mask_hash(decoded['hash']),
+        }
+
+    def must_update(self, encoded):
+        decoded = self.decode(encoded)
+        return (
+            decoded['work_factor'] != self.work_factor or
+            decoded['block_size'] != self.block_size or
+            decoded['parallelism'] != self.parallelism
```

Comment:
`self._test_scrypt_upgrade('parallelism', 'parallelism', 2)` fails, it seems we shouldn't take `parallelism` into account.

Reply:
We confirmed that `parallelism` should be taken into account.